h/static/scripts/annotator/plugin/pdf.coffee

```diff
@@ -1,302 +1,113 @@
-        # Now we have PDF metadata."
-        resolve()
-      else
-        # PDF metadata is not yet available; postponing extraction.
-        setTimeout ( =>
-          # let's try again if we have PDF metadata.
-          tryIt resolve
-        ), 100
-
-    # Return a promise
-    new Promise (resolve, reject) =>
-      if PDFTextMapper.applicable()
-        tryIt resolve
+  uri: ->
+    @documentLoaded.then ->
+      PDFViewerApplication.url
+
+  getMetadata: ->
+    @documentLoaded.then ->
+      info = PDFViewerApplication.documentInfo
+      metadata = PDFViewerApplication.metadata
+
+      # Taken from PDFViewerApplication#load
+      if metadata?.has('dc:title') and metadata.get('dc:title') isnt 'Untitled'
+        title = metadata.get('dc:title')
+      else if info?['Title']
+        title = info['Title']
       else
-        reject "Not a PDF.js document"
-
-  # Extract the text from the PDF
-  scan: ->
-    # Return a promise
-    new Promise (resolve, reject) =>
-      @_pendingScanResolve = resolve
-      @waitForInit().then =>
-
-        # Initialize our main page data array
-        @pageInfo = []
-
-        # Start the text extraction
-        @_extractPageText 0
-
-  # Manually extract the text from the PDF document.
-  # This workaround is here to avoid depending PDFFindController's
-  # own text extraction routines, which sometimes fail to add
-  # adequate spacing.
-  _extractPageText: (pageIndex) ->
-    # Wait for the page to load
-    @_app.pdfDocument.getPage(pageIndex + 1).then (page) =>
-
-      # Wait for the data to be extracted
-      page.getTextContent().then (data) =>
+        title = document.title
 
-        # First, join all the pieces from the bidiTexts
-        rawContent = (text.str for text in data.items).join " "
+      # This is an experimental URN,
+      # as per http://tools.ietf.org/html/rfc3406#section-3.0
+      urn = "urn:x-pdf:" + PDFViewerApplication.documentFingerprint
+      link = [{href: urn}, {href: PDFViewerApplication.url}]
 
-        # Do some post-processing
-        content = @_parseExtractedText rawContent
+      return {title, link}
 
-        # Save the extracted content to our page information registery
-        @pageInfo[pageIndex] =
-          index: pageIndex
-          content: content
+  update: ->
+    {annotator, pdfViewer} = this
 
-        if pageIndex is @getPageCount() - 1
-          @_finishScan()
-        else
-          @_extractPageText pageIndex + 1
+    stableAnchors = []
+    pendingAnchors = []
+    refreshAnnotations = []
 
-  # This is called when scanning is finished
-  _finishScan: =>
-    # Do some besic calculations with the content
-    @_onHavePageContents()
+    for page in pdfViewer.pages when page.textLayer?.renderingDone
+      div = page.div ? page.el
+      placeholder = div.getElementsByClassName('annotator-placeholder')[0]
 
-    # OK, we are ready to rock.
-    @_pendingScanResolve()
+      switch page.renderingState
```

Comment:
This probably needs explanations. It's interacting with the `anchoring/pdf` code. This code here is ensuring that when pages render/unrender that we're making sure that's detectable later by re-initializing the `textLayer` variable and yanking out the placeholder div that holds the highlight spans for un-rendered pages. Pulling the highlights out causes the annotations to re-anchor below.